java/ql/src/experimental/Security/CWE/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,143 @@
+import DataFlow::PathGraph
+
+/** The Java class `java.security.MessageDigest`. */
+class MessageDigest extends RefType {
+  MessageDigest() { this.hasQualifiedName("java.security", "MessageDigest") }
+}
+
+/** The method call `MessageDigest.getInstance(...)` */
+class MDConstructor extends StaticMethodAccess {
+  MDConstructor() {
+    exists(Method m | m = this.getMethod() |
+      m.getDeclaringType() instanceof MessageDigest and
+      m.hasName("getInstance")
+    )
+  }
+}
+
+/** The method `digest()` declared in `java.security.MessageDigest`. */
+class MDDigestMethod extends Method {
+  MDDigestMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("digest")
+  }
+}
+
+/** The method `update()` declared in `java.security.MessageDigest`. */
+class MDUpdateMethod extends Method {
+  MDUpdateMethod() {
+    this.getDeclaringType() instanceof MessageDigest and
+    this.hasName("update")
+  }
+}
+
+/** The hashing method that could taint the input. */
+class MDHashMethodAccess extends MethodAccess {
+  MDHashMethodAccess() {
+    (
+      this.getMethod() instanceof MDDigestMethod or
+      this.getMethod() instanceof MDUpdateMethod
+    ) and
+    this.getNumArgument() != 0
+  }
+}
+
+/** Gets a regular expression for matching common names of variables that indicate the value being held is a password. */
+string getPasswordRegex() { result = "(?i).*pass(wd|word|code|phrase).*" }
+
+/** Finds variables that hold password information judging by their names. */
+class PasswordVarExpr extends VarAccess {
+  PasswordVarExpr() {
+    exists(string name | name = this.getVariable().getName().toLowerCase() |
+      name.regexpMatch(getPasswordRegex()) and not name.matches("%hash%") // Exclude variable names such as `passwordHash` since their values were already hashed
+    )
+  }
+}
+
+/** Holds if `Expr` e is a direct or indirect operand of `ae`. */
+predicate hasAddExprAncestor(AddExpr ae, Expr e) { ae.getAnOperand+() = e }
+
+/** Holds if `MDHashMethodAccess ma` is a second `MDHashMethodAccess` call by the same object. */
+predicate hasAnotherHashCall(MDHashMethodAccess ma) {
+  exists(MDHashMethodAccess ma2 |
+    ma2 != ma and
+    ma2.getQualifier() = ma.getQualifier().(VarAccess).getVariable().getAnAccess()
+  )
+}
+
+/** Holds if `MethodAccess` ma is a hashing call without a sibling node making another hashing call. */
+predicate isSingleHashMethodCall(MDHashMethodAccess ma) { not hasAnotherHashCall(ma) }
+
+/** Holds if `MethodAccess` ma is invoked by `MethodAccess` ma2 either directly or indirectly. */
+predicate hasParentCall(MethodAccess ma2, MethodAccess ma) { ma.getCaller() = ma2.getMethod() }
+
+/** Holds if `MethodAccess` is a single hashing call that is not invoked by a wrapper method. */
+predicate isSink(MethodAccess ma) {
+  isSingleHashMethodCall(ma) and
+  not hasParentCall(_, ma) // Not invoked by a wrapper method which could invoke MDHashMethod in another call stack. This reduces FPs.
```

Comment:
This restriction is silly. It removes all results in methods that are being called from somewhere - that's removing a lot! And I don't see how that restriction is in any way related to the result being a FP or not. I think this restriction needs to be removed, and if that then produces too many FPs, then I think it's back to the drawing board. As-is I would have to reject the PR.

Reply:
Hi @aschackmull,

Thanks for reviewing this PR. I believe this query is a valid one that handles an important scenario, and it is also doable especially with the help from both you and @smowton. 

In the new version I just committed, I restricted the sink to be the top of the call hierarchy that invokes a hashing method. It renders expected results for all the test cases and test projects I tested so I believe it's ready for review.

Please re-review. Thanks.